### ISSUE TITLE
Type Narrowing Does not Propogate to Tuple From Item

### ISSUE THREAD

**Issue by Beinsezii** [microsoft/pyright#10591]
**Describe the bug**
If you narrow a tuple field such as `assert isinstance(my_tuple[0], int)`, pyright correctly narrows the item itself when accessing from index like `my_tuple[0].conjugate()` but not on the container tuple type `my_tuple` itself.

**Code or Screenshots**
```py
def two_returns() -> tuple[str | bool, int]:
    return "yes", 4


def strict() -> tuple[str, int]:
    two = two_returns()
    assert isinstance(two[0], str)
    two[0].capitalize()  # the actual item is narrowed, this is ok
    return two  # but not the tuple itself, this is err
```
since tuples are immutable I would expect the return soundness here to be guaranteed

**VS Code extension or command-line**
```
> pyright --version
pyright 1.1.401
> pyright test.py
/tmp/test.py:9:12 - error: Type "tuple[str | bool, int]" is not assignable to return type "tuple[str, int]"
```

**Comment by D1-3105** [comment]
+1

**Comment by erictraut** [comment]
Yes, narrowing of a tuple type based on the narrowed type of one of its elements is not supported. For a full list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

We do occasionally add support for new type guard patterns, but each one requires custom logic, adds maintenance costs, and potentially affects overall analysis performance, so the bar is relatively high. For comparison, no other Python type checkers support this narrowing pattern either, so that's a pretty good indication that it's not a commonly-used pattern.

I'm going to reject this enhancement request for now. I'm open to reconsidering it in the future if there is sufficient signal from pyright users that this is an important case to handle. In the meantime, you may be able to get the effect you're looking for by creating a user-defined type guard function (using `TypeGuard` or `TypeIs`).
